File: ds-practice/LinkedList.cpp
#include <iostream>
using namespace std;

class Node{
public:
    int data;
    Node* next;
    Node(int val){
        data = val;
        next = nullptr;
    } 
};

class LinkedList{
    private:
        Node* head;

    public:
        LinkedList() {
            head = nullptr;
        }

        void insertAtEnd(int value){
            Node* newNode = new Node(value);
            if (head == nullptr){
                head = newNode;
                return;
            }
            Node* current = head;
            while (current->next){
                current = current->next;
            }
            current->next = newNode;
        }

        void insertAtBeginning(int value){
            Node* newNode = new Node(value);
            newNode->next = head;
            head = newNode;
        }

        void insertAtPosition(int val, int pos){
            Node* newNode = new Node(val);
            if (pos == 0) {
                insertAtBeginning(val);
                return;
            }
            int temp = 1;

            Node* current = head;
            while (temp < pos && current){
                current = current->next;
                temp++;
            } 
            Node* tempNode = current->next;
            current->next = newNode;
            newNode->next = tempNode;
        }

        void print() {
            cout << "List: ";
            Node* current = head;
            while (current) {
                cout << current->data << " ";
                current = current->next;
            }
            cout << endl;
        }

        void reverse(){
            Node* current = head;
            Node* prev = nullptr;
            Node* next_node = nullptr;

            while (current){
                next_node = current->next;
                current->next = prev;
                prev = current;
                current = next_node;
            }

            head = prev;
        }

        void removeFromBeginning() {
            if (head) { head = head->next; }
        }

        void removeFromEnd() {
            if (head) {
                Node* current = head;
                while (current->next->next) {
                    current = current->next;
                }
                current->next = nullptr;
            }
        }

        void removeFromPosition(int pos) {
            if (head) {
                int temp = 1;
                Node* current = head;
                while (temp < pos && current->next->next) {
                    current = current->next;
                    temp++;
                }
                current->next = current->next->next;
            }
        }
};


int main() {
    LinkedList list;
    list.insertAtEnd(1);
    list.insertAtEnd(2);
    list.insertAtEnd(3);

    list.print();

    list.insertAtBeginning(5);
    list.insertAtBeginning(4);

    list.print();

    list.insertAtPosition(7, 3);    
    list.print();

    list.removeFromBeginning();
    list.removeFromEnd();
    list.print();

    list.removeFromPosition(2);
    list.print();

    list.insertAtPosition(7, 2);
    list.print();
    // list.reverse();
    // list.print();
}

File: ds-practice/LinkedList.py
class Node:
    def __init__(self, val):
        self.data = val
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def insert_at_end(self, val):
        newNode = Node(val)
        if self.head is None:
            self.head = newNode
            return
        
        current = self.head
        while current.next:
            current = current.next

        current.next = newNode

    def insert_at_beginning(self, val):
        newNode = Node(val)
        newNode.next = self.head
        self.head = newNode

    def insert_at_position(self, val, pos):
        newNode = Node(val)
        if pos == 0:
            self.insert_at_beginning(val)
            return
        
        temp = 1
        current = self.head
        
        while (temp < pos and current):
            temp += 1
            current = current.next

        tempNode = current.next
        current.next = newNode
        newNode.next = tempNode

    def __repr__(self):
        current = self.head
        res = "List: "

        while current:
            res += str(current.data) + " "
            current = current.next

        return res
        

    def reverse(self):
        current = self.head
        nextNode = None
        prev = None

        while current:
            nextNode = current.next
            current.next = prev
            prev = current
            current = nextNode

        self.head = prev

    def remove_from_beginning(self):
        if self.head:
            head = head.next
        
        return
    
    def remove_from_end(self):
        if self.head:
            current = self.head
            
            while current.next.next:
                current = current.next

            current.next = None

        return
    
    def remove_from_position(self, pos):
        if self.head:
            current = self.head
            temp = 1

            while temp < pos and current.next.next:
                current = current.next
                temp += 1

            current.next = current.next.next

        return
    

ll = LinkedList()
for i in range(1, 11):
    ll.insert_at_end(i)

print(ll)

ll.reverse()
print(ll)

ll.insert_at_position(99, 0)
print(ll)

ll.remove_from_position(7)
print(ll)